feat: Discord rich presence

ryanzhoudev opened the pull request.

ryanzhoudev committed:
feat: Discord rich presence

ryanzhoudev committed:
Merge branch 'main' into discordtest

Commit:
ci: spotless formatting

ryanzhoudev committed:
fix a lot of random issues

kristofbolyai commented on common/src/main/java/com/wynntils/mc/mixin/LocalPlayerMixin.java:
```diff
@@ -49,4 +50,23 @@ private void playSoundPre(SoundEvent sound, float volume, float pitch, CallbackI
+                        target = "(DDDFFZ)Lnet/minecraft/network/protocol/game/ServerboundMovePlayerPacket$PosRot;"),
+                @At(
+                        value = "NEW",
+                        target = "(DDDFF)Lnet/minecraft/network/protocol/game/ServerboundMovePlayerPacket$Pos;"),
```
I am wondering how you got this signature. I checked and the constructor doesn't even have that many variables.

 ```suggestion
                         target = "(DDDZ)Lnet/minecraft/network/protocol/game/ServerboundMovePlayerPacket$Pos;"),
```

Can you confirm which one is correct?

ryanzhoudev replied:
Probably copilot and I didn't notice since the first one gets tripped a lot anyway. DDDZ is correct.

kristofbolyai commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,101 @@
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class DiscordRichPresenceFeature extends Feature {
+
```
```suggestion
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,101 @@
+    @SubscribeEvent
+    public void onCharacterUpdate(CharacterUpdateEvent event) {
+        if (displayCharacterInfo.get()) {
+            Managers.Discord.setClassType(Models.Character.getClassType());
```
IMO this is a little weird connection between managers and functions, perhaps @magicus can suggest a better way when he gets to review this.

magicus replied:
You mean between managers and features, I presume. 

Well, maybe yes. I think there are two reasonable splits to make here, and the current solutions blends those together.

Option 1 is to have a DiscordManager/DiscordExternalModel (or whatever we should call it) that is basically a thin wrapper around the discord API. In that case, it should not know about concepts like "Location" but rather "Large Text" (or whatever it was called).

Option 2 is to have the DiscordManager/DiscordExternalModel specify in detail how we are allowed to set rich presence info on discord, and deal in terms of "Location" and "Class" etc. But in that case, the model should itself listen to the events needed to keep it up to date, which renders the Feature almost moot. At this point, it would only be there to store config and have a way to turn this on/off; since we can't do that for Models yet.

I'm not sure which I prefer. But I think Option 1 have some kind of appeal. That would make the DiscordManager expose a more high level API with text and images (I guess?) to display, and the responsibility of setting those to sane Wynncraft values should fall on this feature. I think that would be neat.

kristofbolyai commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,101 @@
+                Managers.Discord.setLocation("");
+                Managers.Discord.setWorld("In Hub");
+                Managers.Discord.setWynncraftLogo();
+            }
```
Use a switch case please

kristofbolyai commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,101 @@
+            }
+            Managers.Discord.setClassType(null);
+            Managers.Discord.setWynncraftLogo();
+        }
```
This should just fall under the display location config and into the switch

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public void clearAll() {
+        core.activityManager().clearActivity();
```
Perhaps we should free the activity here and create a new one if needed. `activity.close()` and set it to null.

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,92 @@
+            activity.timestamps().setStart(Instant.now());
+        } catch (RuntimeException e) {
+            WynntilsMod.error("Could not initialize Discord Game SDK", e);
+        }
```
I don't think this should run here. Let's add some dependency tree in here, basically like `ChatHandler`'s `dialogExtractionDependents`. I want to give the user the ability to completely bypass running Discord SDK. With this said, you need to move this to an enable method, and create a disable method where you need to clean up all your `AutoCloseable`s.

Lastly, I think you should widen the catch to `Throwable` however I don't think we can catch SDK loading errors.

kristofbolyai reviewed (commented):
Sorry for the many comments, I did not get to tell you how I imagined the manager to work yesterday. Overall it is really close and you just need to refactor it so it can be "turned off".

As for testing. Can you let me know what platforms you tested on? I want to test 2 things on a platform. Running the game with Discord open **and with Discord closed**. I heard some issues back in Wynntils with running the SDK with a closed Discord, let's make sure it doesn't happen here.

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,92 @@
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class DiscordManager extends Manager {
+
```
```suggestion
```

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,92 @@
+        DiscordGameSDKCore.loadLibrary();
+        CreateParams params = new CreateParams();
+        try {
+            params.setClientID(387266678607577088L);
```
Can you perhaps move this into a const?

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,92 @@
+        super(List.of());
+
+        DiscordGameSDKCore.loadLibrary();
+        CreateParams params = new CreateParams();
```
Oh, I just noticed. You are leaking memory here with `params`. It is auto closable, but you mustn't close it until you close core. Which means you need to store this as a field too. I am thinking if you should have a holder record for all these so you don't forget to close all of them.

ryanzhoudev committed:
clearActivity workaround

ryanzhoudev committed:
spotless

ryanzhoudev committed:
DDDFF -> DDDZ

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

ryanzhoudev committed:
remove blank line

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

ryanzhoudev committed:
use switch

ryanzhoudev committed:
loading/unloading

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    public void load() {
+        if (!libraryLoaded) {
```
Comments mention that library can be reloaded so I am not sure if we should complicate this here.

ryanzhoudev committed:
clean up unloading

magicus commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
I'd rather see this in `core/net/discord`, analogous to `core/net/athena` (and setting the pattern for any future net services).

magicus replied:
Actually, this is borderline Manager stuff... It is part of the small but tricky-to-handle set of stuff that is neither "essential to Wynntils operations" (which characterize Managers) nor "a conceptual part of the Wynncraft world" (which characterize Models). It is more like an `ExternalModel` or something like that. 

Other stuff that fall into this category is Hades (which is considered a `Model`) and the Splashes (which is considered a `Manager`).

kristofbolyai replied:
This is so much what I suggested, but you have not got around checking discord :). I even linked the original discussing about this, when I made PingModel.

magicus replied:
I think (no promises!) that I actually will have some time now to do a bit working on Artemis! :) I'll check up on Discord once I've finished handling the Github notifications.

magicus commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,121 @@
+ */
+package com.wynntils.core.discord;
+
+import com.wynntils.antiope.core.DiscordGameSDKCore;
```
Is this weird package name due to shadowing? How come we do not see that for other 3rd party libraries?

kristofbolyai replied:
Not sure what you mean?

magicus replied:
That an external library gets stuck under `com.wynntils`. But I see we have the same when shadowing the Hades lib, so I guess I'll just have to live with it. (If it is possible to make shadowing land in e.g. `com.wynntils.shadow.*` instead of just `com.wynntils`, I'd prefer that, but it is separate from this PR.

kristofbolyai replied:
It is in com.wynntils since we made the package (from a fork)

magicus replied:
Aha, I see. I missed that part. Then it all makes sense.

magicus commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,121 @@
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class DiscordManager extends Manager {
+    private static final long DISCORD_APPLICATION_ID = 387266678607577088L;
```
What is this magic number? Is it like a API key, something we have registered with Discord?

magicus replied:
In that case, I think it is a bit disturbing to hard-code it in the source. I wonder if there is a way that we can read it from a suitable json file instead. urls.json springs to mind, but that only contains URLs. Hm. I need to think a bit about this.

DevScyu replied:
This Client ID should be fine to hard code, as long as it is our application that we have already setup, which it looks to be. This value should never change unless discord does something funky

magicus commented on common/src/main/java/com/wynntils/mc/mixin/LocalPlayerMixin.java:
```diff
@@ -49,4 +50,23 @@ private void playSoundPre(SoundEvent sound, float volume, float pitch, CallbackI
+            },
+            cancellable = true)
+    private void sendPositionPre(CallbackInfo ci) {
+        PlayerMoveEvent event = new PlayerMoveEvent();
```
This strikes me a bit as odd, that we have a event for moving that does not even include the new location we're moving to.

I wonder if we really need to do this. Afaict, this is used to update the Territory info in Discord. I think we might just as well check the current territory every second or so, and update Discord if it has changed. This looks like it add unnecessary overhead in a lag-sensitive part of the game.

magicus commented:
Also, @DonkeyBlaster, thanks for implementing this! That was one of the remaining major functionality outages from Legacy in Artemis. Overall I think this is good, we just need to find the best way to fit it in the Artemis design.

ryanzhoudev committed:
don't complicate loading

ryanzhoudev committed:
refactor manager/feature

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev committed:
position null check

ryanzhoudev committed:
weaken to position

ryanzhoudev commented:
> As for testing. Can you let me know what platforms you tested on? I want to test 2 things on a platform. Running the game with Discord open **and with Discord closed**. I heard some issues back in Wynntils with running the SDK with a closed Discord, let's make sure it doesn't happen here.

On Windows, launching the game without Discord open just opens Discord. Same thing as Legacy. Should probably find a way to prevent this.

ryanzhoudev committed:
Merge branch 'main' into discordtest

kristofbolyai commented:
Can I ask to not implement this while we have the beta and pre-alpha stuff going on? Once we sort out the Wynn update, we are going to come back to this and implement this, it has been missing for too long now.

ryanzhoudev committed:
Merge branch 'main' into discordtest

kristofbolyai commented:
> > As for testing. Can you let me know what platforms you tested on? I want to test 2 things on a platform. Running the game with Discord open **and with Discord closed**. I heard some issues back in Wynntils with running the SDK with a closed Discord, let's make sure it doesn't happen here.
> 
> On Windows, launching the game without Discord open just opens Discord. Same thing as Legacy. Should probably find a way to prevent this.

Just to note this, it is easy to do this. You have to set CreateParams to not require Discord open, and if an exception happens in the callback loop, you know Discord is not opened. This case, we can just close the whole thing and implement a `/wynntils reannouncerichpresence` (or smth) command

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev committed:
bump antiope version

magicus committed:
feat: Integrate 2.0.3 beta changes from development into main (#1722)

* feat: Add a feature to warn users to use pre-alpha for beta wynncraft (#1653)

Co-authored-by: Scyu_ <scyu@scyu.dev>

* fix: fix snapshot jar name

* fix: Fix lootrun scoreboard part not showing up (#1656)

* fix: fix version check (#1657)

* fix: 🐛fix stream version check

* ci: spotless formatting

---------

Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>

* fix: 💚fix version replacer

* fix: Support admin rank in login redirect (#1660)

* fix: Fix shout recipient type parsing (#1661)

* fix: Fix NEW stats being calculated in overall quality (and displaying as perfect) (#1664)

* fix: Do not crash on chat encoding for mismatched stats (#1669)

* fix: Beta/pre-alpha usability fixes (#1667)

fix: Add option to connect to beta from Wynncraft button.
fix: Hard-code development version to behave as pre-alpha.
fix: On pre-alpha, warn if user connects to non-beta server.

* [ci skip] chore: Add comment about our filtering policy (#1673)

* fix: Make shiny items parse as gear (#1677)

* refactor: Make party leader in PartyModel safer to use (#1675)

* refactor: Make party leader in PartyModel safer to use

* refactor: Introduce McUtils.playerName()

* feat: Content book scrolling (#1678)

* feat: Show queue position in loading screen (#1679)

feat: Show Queue position in loading screen

* fix: Add fallback handler to scoreboard handler (#1671)

* fix: Add fallback handler to scoreboard handler

Otherwise new, unknown segments are just lost

* ci: spotless formatting

* Delete the placeholder lootrun scoreboard part

* ci: spotless formatting

* Delete the placeholder party finder scoreboard part

---------

Co-authored-by: magicus <magicus@users.noreply.github.com>

* feat: New Tracker Model, separated from QuestModel (#1672)

* fix: beta: Update matcher to new GUI title

* fix: beta: Update tracked name matcher

* refactor: Convert most of the "QuestInfo" tracking to a general "tracker" model.

Also start adding some more general tracking functionality.

* fix: Also update resource location in source

* feat: Split Tracker and Quest model properly

* Misc cleanup

* Rename Tracker to ContentTracker wherever feasible

* ci: spotless formatting

* Always calls matches(). Docs is a bit unclear if this is strictly needed or not, but let's play safely.

* Also rename i18n...

* Even more i18n...

* Add config upfixer

* Rename upfixer

---------

Co-authored-by: magicus <magicus@users.noreply.github.com>

* fix: Fix auto update not working on any stream (#1680)

* feat: Parse Content Book items (#1683)

* feat: Create ContentModel and ContentItem

* feat: Improve item debug tooltips

* spotless

* Add parsing of most of the content lore

* Fix remaining parsing

* Give examples of specialInfo.

* Use INCLUDE_EVENTS for toString in StyledText

* Clarify how we detect when a req is fulfilled

* Remove hardcoded color values from content status

* I'm only changing this since it aligns better with the change on the line below...

* Move normalization to when joinLines is called

* Improve null safety

* refactor: Merge ContentModel and ContentTrackerModel (#1684)

* feat: Create ContentModel and ContentItem

* feat: Improve item debug tooltips

* spotless

* Add parsing of most of the content lore

* Fix remaining parsing

* refactor: Move contenttracker in as part of content

* refactor: Merge ContentModel with ContentTrackerModel

* Give examples of specialInfo.

* Use INCLUDE_EVENTS for toString in StyledText

* Clarify how we detect when a req is fulfilled

* Remove hardcoded color values from content status

* I'm only changing this since it aligns better with the change on the line below...

* Move normalization to when joinLines is called

* Improve null safety

* fix: i18n problems with content tracker title etc (#1686)

* fix: i18n problems with content tracker title etc

* ci: spotless formatting

* Fix merge error

---------

Co-authored-by: magicus <magicus@users.noreply.github.com>

* fix: Container Queries did not get properly annotated with WynnItems (#1689)

* fix: Container Queries did not get properly annotated with WynnItems

* ci: spotless formatting

---------

Co-authored-by: magicus <magicus@users.noreply.github.com>

* fix: Fix content tracker overlay parsing being broken (#1693)

* fix: Fix info bar not matching (#1695)

* fix: Proper fix for duplicated container content events (#1696)

fix: Proper fix for duplicated container content events when using the container query

* feat: Refactor WynncraftButtonFeature connectTo* configs to an enum (#1694)

* chore: Refactor WynncraftButtonFeature connectTo* configs to an enum

* fix: use correct address

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* fix: Parse loot chests correctly (#1698)

* feat: Add support for new Dialogue History (#1705)

Also add expectSameContainer for container queries

* Merge from main (#1699)

* feat: Add a feature to warn users to use pre-alpha for beta wynncraft… (#1655)

feat: Add a feature to warn users to use pre-alpha for beta wynncraft (#1653)

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* fix: Fix map moving when left-clicking (#1632)

* fix: Fix map moving when left-clicking

* make use of ContainerEventHandler#mouseDragged

* refactor: Add braces

* fix: Changing class does not clear status overlay (#1663)

* feat: Option to change emerald count text display side (#1658)

* Add an option to change emerald count text display side

* Some small fixes

* Some final small fixes

* ci: spotless formatting

* chore: remove empty line

* fix: simplify i18n

---------

Co-authored-by: jh-devv <jh-devv@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* feat: Highlight search when reopening screen (#1649)

* feat: Highlight search when reopening screen

* use Screen#added instead of Screen#onClose

* fix: fix updater and version checking (#1665)

* fix: fix snapshot jar name

* fix: fix version check (#1657)

* fix: 🐛fix stream version check

* ci: spotless formatting

---------

Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>

* fix: 💚fix version replacer

---------

Co-authored-by: Scyu_ <scyu@scyu.dev>
Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>

* [ci skip] chore: Speed up build by caching (correctly) (#1666)

Co-authored-by: Scyu_ <scyu@scyu.dev>

* fix: LoadingScreen doesn't show up when joining world (#1628)

* fix: LoadingScreen doesn't show up when loading world

* remove ScreenOpenedEvent.AlwaysPre

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* fix: Fix auto update not working on any stream (#1681)

* Merge from main

* fix: Fix function parsing handling newline (and other cases) weirdly + fix /function test output (#1701)

fix: Fix function parsing handling newline (and other cases) weirdly + fix /function test

* feat: Add an option to override item favorite blocker (#1700)

* feat: Add an option to override item favorite blocker

* fix: Flip comparison

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* fix: Function names can contain '_' (#1703)

* feat: Add feature descriptions in config screen (#1704)

* feat: Add description to i18n

* feat: Render it

* chore: no need for this

* fix: Fix i18n

---------

Co-authored-by: Scyu_ <scyu@scyu.dev>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>
Co-authored-by: Martin <66840021+Mcrtin@users.noreply.github.com>
Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>
Co-authored-by: jh-devv <122896463+jh-devv@users.noreply.github.com>
Co-authored-by: jh-devv <jh-devv@users.noreply.github.com>
Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>

* chore: Rebuild with changes from main

* Merge from main into development (#1712)

* feat: Add a feature to warn users to use pre-alpha for beta wynncraft… (#1655)

feat: Add a feature to warn users to use pre-alpha for beta wynncraft (#1653)

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* fix: Fix map moving when left-clicking (#1632)

* fix: Fix map moving when left-clicking

* make use of ContainerEventHandler#mouseDragged

* refactor: Add braces

* fix: Changing class does not clear status overlay (#1663)

* feat: Option to change emerald count text display side (#1658)

* Add an option to change emerald count text display side

* Some small fixes

* Some final small fixes

* ci: spotless formatting

* chore: remove empty line

* fix: simplify i18n

---------

Co-authored-by: jh-devv <jh-devv@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* feat: Highlight search when reopening screen (#1649)

* feat: Highlight search when reopening screen

* use Screen#added instead of Screen#onClose

* fix: fix updater and version checking (#1665)

* fix: fix snapshot jar name

* fix: fix version check (#1657)

* fix: 🐛fix stream version check

* ci: spotless formatting

---------

Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>

* fix: 💚fix version replacer

---------

Co-authored-by: Scyu_ <scyu@scyu.dev>
Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>

* [ci skip] chore: Speed up build by caching (correctly) (#1666)

Co-authored-by: Scyu_ <scyu@scyu.dev>

* fix: LoadingScreen doesn't show up when joining world (#1628)

* fix: LoadingScreen doesn't show up when loading world

* remove ScreenOpenedEvent.AlwaysPre

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* fix: Fix auto update not working on any stream (#1681)

* fix: Fix function parsing handling newline (and other cases) weirdly + fix /function test output (#1701)

fix: Fix function parsing handling newline (and other cases) weirdly + fix /function test

* feat: Add an option to override item favorite blocker (#1700)

* feat: Add an option to override item favorite blocker

* fix: Flip comparison

---------

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* fix: Function names can contain '_' (#1703)

* feat: Add feature descriptions in config screen (#1704)

* feat: Add description to i18n

* feat: Render it

* chore: no need for this

* fix: TM filter and search did not work with scroll (#1702)

* fix: TM filter and search did not work with scroll

* use regex

* fix: Fix objective scoreboard part showing incorrectly when either objective overlay is disabled (#1707)

* fix: Fix changelog showing when first installing, bad changelog spacing, and add scrolling (#1706)

* feat: Add guild name and rank to book above player (#1652)

* feat: Add guild name to book above player

* ci: spotless formatting

* Added guild rank and retrieves info from compass menu now, not boss bar

* ci: spotless formatting

* Removed prints

* Added rank stars, moved out of CharacterModel into GuildModel

* Depends on CharacterModel and now uses GuildRank

* Added stars to enum

* Replace getStars with getGuildDescription

* Update common/src/main/java/com/wynntils/models/players/type/GuildRank.java

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* Update common/src/main/java/com/wynntils/models/players/type/GuildRank.java

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* Update common/src/main/java/com/wynntils/models/players/type/GuildRank.java

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* Update common/src/main/java/com/wynntils/screens/wynntilsmenu/WynntilsMenuScreen.java

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* Use name not name()

---------

Co-authored-by: JamieCallan117 <JamieCallan117@users.noreply.github.com>
Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* feat: render lootrun path on map (#1688)

* feat: render lootrun path on map

* feat: render lootrun path on map

* Update common/src/main/java/com/wynntils/utils/render/MapRenderer.java

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* offload method to utils class

* merge render calls, changed up the render order slightly to better portrait line continuity

* simplification optimization, merge the point filtering process

---------

Co-authored-by: Halcyon <@>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* feat: Silencer - hotkey-toggleable assignable Master Volume (#1643)

* feat: Silencer - hotkey-toggleable assignable Master Volume

* fix: missing ConfigCategory

* refactor: silencerVolume description

Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>

* refactor: alphabetical order

* fix: +reliability by use onTitleScreenInit

* fix: reset no longer triggers every time onTitleScreenInit is called.
refactor: improved naming and added documentation.

* fix: forgot check for masterVolume == silencerVolume

* refactor: comment in onTitleScreenInit for restoreOriginalVolume

Co-authored-by: Magnus Ihse Bursie <mag@icus.se>

* ci: spotless formatting

---------

Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>
Co-authored-by: Magnus Ihse Bursie <mag@icus.se>
Co-authored-by: magicus <magicus@users.noreply.github.com>

* fix: NPE in upfixer (#1710)

---------

Co-authored-by: Scyu_ <scyu@scyu.dev>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>
Co-authored-by: Martin <66840021+Mcrtin@users.noreply.github.com>
Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>
Co-authored-by: jh-devv <122896463+jh-devv@users.noreply.github.com>
Co-authored-by: jh-devv <jh-devv@users.noreply.github.com>
Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>
Co-authored-by: Jamie <JamCal117@gmail.com>
Co-authored-by: JamieCallan117 <JamieCallan117@users.noreply.github.com>
Co-authored-by: HalcyonSuoh <60712627+HalcyonSuoh@users.noreply.github.com>
Co-authored-by: re-vc <73580835+re-vc@users.noreply.github.com>
Co-authored-by: magicus <magicus@users.noreply.github.com>

* fix: typo (#1719)

* fix: Fix the broken fix for loading screen titles (#1720)

* feat: Add support for the new lootrun camp content type (#1721)

* refactor: Use named colors instead of hard-coded color codes (#1723)

* [ci skip] feat: Redesign Container Queries to allow dynamic queries, and support the new Content Book (#1697)

* WIP: feat: Read new quest book and turn it into our old QuestInfo

* Also fix our new content book

* Spotless

* Remove acceptNoOp

* Support conversion between QuestInfo and ContentInfo

* Disable old quest and discovery queries which are not needed

* Implement first iteration of new Container Query API

* WIP: Partial progress. Does not compile!

* Remove onComplete

* Extract simple types

* spotless

* Sort of might work now...

* Remove some old code

* Fix AbilityTreeContainerQueries properly

* bug fix

* Fix most of content/quest book, except filters

* Ugly temp fix to handle that next step might not have anything to do, and this is not necessarily a bug.

* Some cleanups

* Even more cleanups

* Improve content book callback

* Track mini-quests

* Fix Discoveries, mostly at least

* Fix filter!

* Fixup discoveries

* Silence click sound

* Fix documentation

* Remake API so we throw exceptions on errors instead

* Remove printf

* Cleanups after suggestion by IntelliJ

* More suggestions by IntelliJ

* Fix tooltips

* Fix tracking for mini-quests

* ci: spotless formatting

* Include error message

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

---------

Co-authored-by: magicus <magicus@users.noreply.github.com>
Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

* [ci skip] feat: Replace the new content book (#1724)

---------

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>
Co-authored-by: Scyu_ <scyu@scyu.dev>
Co-authored-by: DevScyu <DevScyu@users.noreply.github.com>
Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>
Co-authored-by: magicus <magicus@users.noreply.github.com>
Co-authored-by: Martin <66840021+Mcrtin@users.noreply.github.com>
Co-authored-by: jh-devv <122896463+jh-devv@users.noreply.github.com>
Co-authored-by: jh-devv <jh-devv@users.noreply.github.com>
Co-authored-by: Jamie <JamCal117@gmail.com>
Co-authored-by: JamieCallan117 <JamieCallan117@users.noreply.github.com>
Co-authored-by: HalcyonSuoh <60712627+HalcyonSuoh@users.noreply.github.com>
Co-authored-by: re-vc <73580835+re-vc@users.noreply.github.com>

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev committed:
Merge remote-tracking branch 'origin/discordtest' into discordtest

ryanzhoudev committed:
fix most rpc issues

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev commented:
Works fine on Windows now, from my testing. Someone please check macOS and Linux.

magicus commented:
I just tested this on my M1 mac, and it works fine!

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,115 @@
+            WynntilsMod.error("Could not initialize Discord Game SDK", e);
+        } catch (Throwable e) {
+            WynntilsMod.error("Could not initialize Discord Game SDK", e);
+        }
```
we could merge these and do an instanceof

magicus replied:
At first I thought you were borderline crazy to suggest using instanceof on exceptions, but now I see what you mean. The special case is if the exception is GameSDKException and the result is INTERNAL_ERROR, then we accept it as okay.

So yes, then I agree.

magicus commented:
I tested that it worked fine in updating my info when I had discord started and then started Minecraft. I also checked that it did not start Discord by itself if it was not running.

However, if I started Discord after I loaded Minecraft, we did not discover this and start sending updates. I think we should. At the very least try to look for discord again each time we switch character or something like that.

kristofbolyai commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,115 @@
+        if (isReady()) {
+            core.runCallbacks();
+        }
+    }
```
can we do smth like hades model does and only tick once like every 5 ticks?

magicus commented:
What happens if you don't even have Discord installed? Have you tested that?

Also, is there some way we can detect if that is the case, and if so, skip trying to re-detect Discord as I suggested above?

kristofbolyai commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
I think this is the singular file blocking this feature.. This feature does way too much. What blocks us from having a simple "updatePresenceInfo" method and that can get called by all the events, etc?

magicus replied:
You think so? I think it does exactly what it needs to, neither too much nor to little. Possibly I like it because @DonkeyBlaster is using the same kind of coding style that I am using :-) but I think it looks okay. What do you mean that the alternative would be? That every change, whatever it is, would trigger us to recalculate the entire discord info?

You would still need to do the WorldStateEvent thing, where you check if you are on the hub etc. And if you get a callback from the territory check every 10th tick or so, it makes no sense to be re-checking if you are on the hub or not every time that happens; instead it makes sense to just update the territory.

So I don't think it would be an improvement. 

Or I misunderstand you..?

ryanzhoudev replied:
If we do the updatePresenceInfo thing, wouldn't we be delegating too much to the manager?

magicus replied:
I'm assuming the updatePresenceInfo() method should be in the Feature.

ryanzhoudev replied:
That is almost what displayCharacterDetails() does then. Currently it just sets it to the stored fields in the feature, presumably this proposed function will just get the details at the time and set it immediately. Currently the correct events set the stored fields.

magicus commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,154 @@
+    private ClassType classType = null;
+    private int level = 0;
+
+    private void setCharacterDetails() {
```
I don't think it is a formal rule, but I prefer to put the `@SubscribeEvent` methods before private methods, and I think I've noticed @kristofbolyai does the same.

magicus commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,154 @@
+            }
+        }
+
+        Managers.TickScheduler.scheduleLater(this::checkTerritory, 10);
```
Please make this value a constant

magicus commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,154 @@
+        if (!Models.WorldState.onWorld()) return;
+
+        // classType needs to be set even when config is disabled so if the config is enabled later, it will not require
+        // a relog
```
Why is that? You are calling setCharacterDetails() from onConfigChange.

ryanzhoudev replied:
setCharacterDetails() doesn't set the feature's classType field, it just tells the rich presence to display what is already stored. This is a misleading name now that you brought it up, I will change it.

magicus replied:
I was more wondering why you need a classType field at all, and not just read it from the model whenever you need to change the class info?

ryanzhoudev replied:
Originally it was so that when the display function was called it wouldn't have to ask the model for the classType as much, but I suppose both are reading some field and that's it, so it would be the same for performance

magicus reviewed (commented):
Overall this looks very good, apart from my remarks/questions.

Thanks for filling in this blank from the legacy partity!

ryanzhoudev commented:
> What happens if you don't even have Discord installed? Have you tested that?
> 
> Also, is there some way we can detect if that is the case, and if so, skip trying to re-detect Discord as I suggested above?

I believe not having Discord installed and not having Discord launched will result in the same state for our code.

magicus commented:
> I believe not having Discord installed and not having Discord launched will result in the same state for our code.

Sounds reasonable, now that I think about it. :) I was confused by the "launch Discord if not already running" part.

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev committed:
merge catch clause

ryanzhoudev committed:
every 5 ticks

Commit:
ci: spotless formatting

magicus commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,118 @@
+
+    @SubscribeEvent
+    public void onTick(TickEvent event) {
+        if (!isReady() || ticksUntilUpdate > 0) return;
```
Maybe do ticksUntilUpdate-- somewhere here..? :-)

Also, please check ticksUntilUpdate first, since it is quickest. isReady() calls `.core.isOpen`. This sounds like it might be expensive; at the very least we should only check it once the TICKS_PER_UPDATE ticks have passed.

ryanzhoudev committed:
move territory delay to const

ryanzhoudev committed:
fix tick logic

ryanzhoudev committed:
change method name

ryanzhoudev committed:
rearrange function order

ryanzhoudev committed:
Merge branch 'main' into discordtest

magicus commented on common/src/main/java/com/wynntils/core/discord/DiscordManager.java:
```diff
@@ -0,0 +1,122 @@
+        if (!isReady()) return;
+
+        core.runCallbacks();
+        ticksUntilUpdate = TICKS_PER_UPDATE;
```
You need to move this up to before the isReady test, otherwise it will be tested every tick if it fails, since ticksUntilUpdate will remain at 0.

ryanzhoudev committed:
fix ticking for real

ryanzhoudev committed:
add small image

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev committed:
don't store in feature

ryanzhoudev committed:
Merge branch 'main' into discordtest

magicus reviewed (approved):
I think this looks great now. If @kristofbolyai strongly disagrees, we'll have to continue discuss how this should be solved. But from my PoV, this is perfectly fine.

magicus committed:
Merge branch 'main' into discordtest

magicus commented on common/src/main/java/com/wynntils/features/DiscordRichPresenceFeature.java:
```diff
@@ -0,0 +1,154 @@
+    public void onXpChange(SetXpEvent event) {
+        if (!Models.WorldState.onWorld()) return;
+
+        // same as above, level needs to be set even when config is disabled
```
This comment seems to be left behind, though...

ryanzhoudev committed:
remove wrong comment

kristofbolyai reviewed (approved):
Okay, I can accept this. Just make sure you merge it in at a time where you have time to deal with potential issues. I've had a long 3 days debugging container queries...

ryanzhoudev committed:
Merge branch 'main' into discordtest

ryanzhoudev commented:
I have time to deal with issues today, tomorrow, and the afternoon of the day after, so I will merge now.
